src/beamlime/communication/pipes.py

```diff
@@ -0,0 +1,238 @@
+        if _chunk_size > MAX_CHUNK_SIZE:
+            raise ValueError("Maximum chunk size is limited to ", MAX_CHUNK_SIZE)
+        self._chunk_size = _chunk_size
+
+    @contextmanager
+    def _open_buffer(
+        self, buffer_type: BufferType[BufferData]
+    ) -> Iterator[Buffer[BufferData]]:
+        """
+        Create a buffer of the ``buffer_type``.
+        """
+        try:
+            from copy import deepcopy
+
+            chunk, self._data = (
+                self._data[: self.chunk_size],
+                self._data[self.chunk_size :],
+            )
+            _buffer = buffer_type(*deepcopy(chunk), max_size=self.chunk_size)
+            yield _buffer
+        except Exception as err:
+            self._data = chunk + self._data
+            raise err
+        else:
+            consumed = max(0, len(chunk) - len(_buffer))
+            self._data = chunk[consumed:] + self._data
+        finally:
+            ...
+
+    @contextmanager
+    def open_readable(
+        self, timeout: int = 0, retry_interval: float = 0.1
+    ) -> Iterator[BufferBase[BufferData]]:
+        """
+        Yield a buffer containing a copy of the first chunk of data.
+        """
+        from ..core.schedulers import retry
+
+        max_trials = int(timeout / retry_interval) + 1
+
+        @retry(Empty, max_trials=max_trials, interval=retry_interval)
+        def wait_for_data() -> None:
+            if not self._data:
+                raise Empty("There is no data left in the pipe.")
+
+        wait_for_data()
+        try:
+            with self._open_buffer(SyncBuffer[BufferData]) as _buffer:
+                yield _buffer
+        finally:
+            ...
+
+    @asynccontextmanager
+    async def open_async_readable(
+        self, timeout: int = 0, retry_interval: float = 0.1
+    ) -> AsyncIterator[BufferBase[BufferData]]:
+        """
+        Yield an async buffer containing a copy of the first chunk of data.
+        """
+
+        from ..core.schedulers import async_retry
+
+        max_trials = int(timeout / retry_interval) + 1
+
+        @async_retry(Empty, max_trials=max_trials, interval=retry_interval)
+        async def wait_for_data() -> None:
+            if not self._data:
+                raise Empty("There is no data left in the pipe.")
+
+        await wait_for_data()
+
+        try:
+            with self._open_buffer(AsyncBuffer[BufferData]) as _buffer:
+                yield _buffer
+        finally:
+            ...
+
+    def write(self, data: BufferData) -> None:
+        """
+        Append a copy of a piece of data into the pipe.
+
+        Raises
+        ------
+        queue.Full
+            If the pipe will have more than ``self.max_size``
+            of data pieces with the new piece appended.
+        """
+        from copy import copy
+
+        if len(self._data) >= self.max_size:
+            raise Full(f"Pipe is full. There are {self.max_size} pieces of data.")
+        self._data.append(copy(data))
```

Comment:
Making a copy? Not sure we can afford this in general. Why is this needed?

Reply:
Yes, you're right... it's not needed.